### PR TITLE
Changes to support `torch._export.aot_compile`

### PR DESCRIPTION
**Pull request by agunapal**
## Description

This PR 
- Adds support for `torch._export.aot_compile`
- Includes an example with ResNet18 with `max-autotune` and `dynamic_shapes`
- Tested equivalence with multiple loads

Comparison with `torch.compile`

<google-sheets-html-origin><style type="text/css"></style>
Model | Mode | Model loading(ms) | First Inference Time (ms)
-- | -- | -- | --
  |   |   |  
ResNet 18 | compile + max-autotune | 4111 | 25918
VGG 16 | compile + max-autotune | 5284 | 64960
ResNet 18 | torch._export.aot_compile + max-autotune | 15876 | 2704
VGG 16 | torch._export.aot_compile + max-autotune | 16042 | 2794

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

```
pytest -v test_torch_export.py 
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.13, pytest-7.3.1, pluggy-1.0.0 -- /home/ubuntu/anaconda3/envs/ts_sam_Dec1/bin/python
cachedir: .pytest_cache
rootdir: /home/ubuntu/serve
plugins: cov-4.1.0, mock-3.12.0
collected 1 item                                                                                                                                                                         

test_torch_export.py::test_torch_export_aot_compile PASSED                                                                                                                         [100%]

==================================================================================== warnings summary ====================================================================================
test_torch_export.py:4
  /home/ubuntu/serve/test/pytest/test_torch_export.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2348
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2348: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(parent)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('ruamel')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/transformers/utils/generic.py:441
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
    _torch_pytree._register_pytree_node(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================= 1 passed, 7 warnings in 45.19s =============================================================================
(ts_sam_Dec1) ubuntu@ip-172-31-11-40:~/serve/test/pytest$ 

```

Checking Equivalance when loading multiple times

Used the following code
```
import pickle
import torch

FILE1 = "test_load.pkl"
FILE2 = "test_load1.pkl"
print("Reading output 1")
with open(FILE1,'rb') as f:
    output1 = pickle.load(f)

print("Reading output 2")
with open(FILE2,'rb') as f:
    output2 = pickle.load(f)


print("Checking if output1 == output2", torch.equal(output1, output2))
```
results in 

```
python test_equivalance.py 
Reading output 1
Reading output 2
Checking if output1 == output2 True

```

- Model Loading logs
```
2023-12-13T20:03:15,259 [INFO ] W-9000-res18-pt2_1.0-stdout MODEL_LOG - torch._export is an experimental feature! Succesfully loaded torch exported model.
2023-12-13T20:03:15,259 [INFO ] W-9000-res18-pt2_1.0-stdout MODEL_LOG - export is not a supported backend
2023-12-13T20:03:15,264 [INFO ] W-9000-res18-pt2_1.0 org.pytorch.serve.wlm.WorkerThread - Backend response time: 17378
2023-12-13T20:03:15,264 [DEBUG] W-9000-res18-pt2_1.0 org.pytorch.serve.wlm.WorkerThread - W-9000-res18-pt2_1.0 State change WORKER_STARTED -> WORKER_MODEL_LOADED
2023-12-13T20:03:15,265 [INFO ] W-9000-res18-pt2_1.0 TS_METRICS - WorkerLoadTime.Milliseconds:19913.0|#WorkerName:W-9000-res18-pt2_1.0,Level:Host|#hostname:ip-172-31-11-40,timestamp:1702497795
2023-12-13T20:03:15,265 [INFO ] W-9000-res18-pt2_1.0 TS_METRICS - WorkerThreadTime.Milliseconds:3.0|#Level:Host|#hostname:ip-172-31-11-40,timestamp:1702497795
```

- Model loaded on 4 GPUs
Screenshot 2023-12-18 at 3 29 31 PM

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by chauhang** [comment]
@ankithagunapal Thanks for adding this support. Why is the model load taking so longer with AOTCompile?

**Comment by agunapal** [comment]
> @ankithagunapal Thanks for adding this support. Why is the model load taking so longer with AOTCompile?

@chauhang Yet to run the profiler on the loading of the .so file. Will check and follow-up

File: examples/pt2/README.md
Comment (msaroufim) [inline examples/pt2/README.md:113]:
```suggestion
Using `torch.compile` to wrap your existing eager PyTorch model can result in out of the box speedups. However, `torch.compile` is a JIT compiler. TorchServe has been supporting `torch.compile` since PyTorch 2.0 release. In a production setting, when you have multiple instances of TorchServe, each of of your instances would `torch.compile` the model on the first inference. TorchServe's model archiver is not able to truly guarantee reproducibility because its a JIT compiler.
```

File: examples/pt2/README.md
Comment (msaroufim) [inline examples/pt2/README.md:117]:
```suggestion
To solve this problem, `torch.export` has an experimental API `torch._export.aot_compile` which is able to `torch.export` a torch compilable model if it has no graph breaks and then run it with AOTInductor
```

File: examples/pt2/torch_export_aot_compile/README.md
Comment (msaroufim) [inline examples/pt2/torch_export_aot_compile/README.md:1]: duplicate docs?

File: examples/pt2/torch_export_aot_compile/README.md
Comment (msaroufim) [inline examples/pt2/torch_export_aot_compile/README.md:21]:
```suggestion
Ex:  `cd  examples/pt2/torch_export_aot_compile`
```

File: examples/pt2/torch_export_aot_compile/README.md
Comment (msaroufim) [inline examples/pt2/torch_export_aot_compile/README.md:25]: we don't need this script, install_dependencies.py has a nightly flag

File: examples/pt2/torch_export_aot_compile/README.md
Comment (msaroufim) [inline examples/pt2/torch_export_aot_compile/README.md:33]: did this interact well with dynamic batching? What happens if you set the batch size to something really large instead?

File: examples/pt2/torch_export_aot_compile/README.md
Comment (msaroufim) [inline examples/pt2/torch_export_aot_compile/README.md:34]: can you bring core idea? Batch size 1 is what people choose if they want low latency

File: requirements/common_gpu.txt
Comment (msaroufim) [inline requirements/common_gpu.txt:3]: is ninja needed to run an inference or to export? If export only we shouldn't add this as a dependency

File: ts/handler_utils/torch_export/load_model.py
Comment (msaroufim) [inline ts/handler_utils/torch_export/load_model.py:11]: would we need to change the name in the case of multiple workers?

File: ts/handler_utils/torch_export/load_model.py
Comment (msaroufim) [inline ts/handler_utils/torch_export/load_model.py:14]: does this use the same cache_dir as teh environment variable a user would be setting?

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:158]: set these global variables in packaging if condition instead

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:56]: i think packaging throws a bunch of warnings right? Can we do whatever the warning recommends cause our logs are starting to get noisy

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:57]: does this mean pt.2.2? Can we call this call PT2_2-0 should be easier to read

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:193]: archiver doc strings need an update too to make it clear it also supports .so file

**Review by msaroufim — CHANGES_REQUESTED** [review]
Looks mostly good, a few minor things left to do - also still not super clear on how the timing measurements were obtained - I'd add some log.info statements

File: ts/handler_utils/torch_export/load_model.py
Comment (agunapal) [reply ts/handler_utils/torch_export/load_model.py:14]: This seems like `TORCHINDUCTOR_CACHE_DIR`

File: requirements/common_gpu.txt
Comment (agunapal) [reply requirements/common_gpu.txt:3]: This is needed for loading the .so file. I can add this in requirements.txt for the example?

Comment (agunapal) [reply requirements/common_gpu.txt:3]: This doesn't work.

**Comment by agunapal** [comment]
> Looks mostly good, a few minor things left to do - also still not super clear on how the timing measurements were obtained - I'd add some log.info statements

@msaroufim Thanks. I have addressed most of the feedback. However, wondering how we should address support for CPU. Does aot_compile on CPU make sense?

File: ts/handler_utils/torch_export/load_model.py
Comment (agunapal) [reply ts/handler_utils/torch_export/load_model.py:11]: tried with multiple workers. Shared a screenshot of nvidia-smi in the PR

File: test/pytest/test_torch_export.py
Comment (agunapal) [reply test/pytest/test_torch_export.py:50]: added. Thanks

File: examples/pt2/torch_export_aot_compile/README.md
Comment (agunapal) [reply examples/pt2/torch_export_aot_compile/README.md]: I read the doc , but I don't understand it tbh

File: examples/pt2/torch_export_aot_compile/README.md
Comment (agunapal) [reply examples/pt2/torch_export_aot_compile/README.md:32]: tried with batch_size 16. It works. I guess we launch it as an experimental feature and then based on issues/feedback, address them

File: examples/pt2/torch_export_aot_compile/README.md
Comment (agunapal) [reply examples/pt2/torch_export_aot_compile/README.md]: resolved. Thanks

File: ts/torch_handler/base_handler.py
Comment (agunapal) [reply ts/torch_handler/base_handler.py]: I looked at the logic again. This seems like the best place.

File: ts/torch_handler/base_handler.py
Comment (agunapal) [reply ts/torch_handler/base_handler.py:56]: I don't see any warnings from packaging

File: ts/torch_handler/base_handler.py
Comment (agunapal) [reply ts/torch_handler/base_handler.py:57]: Doesn't allow `PT2_2-0`

File: ts/torch_handler/base_handler.py
Comment (agunapal) [reply ts/torch_handler/base_handler.py]: not sure which one you are specifically referring to?
This doesn't talk about the format of the file
```
usage: torch-model-archiver [-h] --model-name MODEL_NAME [--serialized-file SERIALIZED_FILE] [--model-file MODEL_FILE] --handler HANDLER [--extra-files EXTRA_FILES]
                            [--runtime {python,python3}] [--export-path EXPORT_PATH] [--archive-format {tgz,no-archive,zip-store,default}] [-f] -v VERSION [-r REQUIREMENTS_FILE]
                            [-c CONFIG_FILE]
torch-model-archiver: error: the following arguments are required: --model-name, --handler, -v/--version
```

**Comment by msaroufim** [comment]
CPU support is a completely valid scenario, inductor codegens native c++ that the intel team has been optimizing

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:162]: why do you need this bool flag?

File: ts/torch_handler/base_handler.py
Comment (msaroufim) [inline ts/torch_handler/base_handler.py:196]: could you add some comment here? Why are we launching things on specific streams?

**Review by msaroufim — APPROVED** [review]
stamping to unblock but this needs a few more things before merge
1.  This should indeed work with cpu so add a test
2. Get rid of the install nightlies script and instead use the good old install_dependencies.py script
3. Add a comment on the cuda streams section
4. You don't need the export bool flag IMO

**Comment by agunapal** [comment]
> stamping to unblock but this needs a few more things before merge
> 
> 1. This should indeed work with cpu so add a test
> 2. Get rid of the install nightlies script and instead use the good old install_dependencies.py script
> 3. Add a comment on the cuda streams section
> 4. You don't need this flag  USE_TORCH_EXPORT

1. Tested on CPU. pytest runs on both CPU/GPU. There is a limitation for batch_size on CPU. Mentioned this in the script.
2. I have updated instructions to have both. Most users using torchserve already have dependencies installed. the torch version doesn't get overwritten. It needs to be uninstalled and then installed again. Hence, script is needed
3. Good catch. Turns out this was not needed. We just need torch.cuda.set_device
4. Refactored the code. 

Also, changed the config to the following. (in case they add other options in export in the future)
```
pt2 :
  export:
    aot_compile: true
```

**Pull request merged**
